File: Java/src/BubbleSort/BubbleSort.java
package BubbleSort;

import sorting.AbstractSort;

public class BubbleSort<T extends Comparable<T>> extends AbstractSort<T> {

	@Override
	public void sort(T[] array, int leftIndex, int rightIndex) {
		if(validateParams(array,leftIndex,rightIndex)){
			//simpleBubbleSort(array,leftIndex,rightIndex);
			//bidirectionalBubbleSort(array, leftIndex, rightIndex);
			recursiveBubbleSort(array, leftIndex, rightIndex, true);
		}else{
			//Do nothing
		}
	}
	
	private void simpleBubbleSort(T[] array,int leftIndex,int rightIndex){
		for(int i = leftIndex ; i < rightIndex;i++){
			boolean flag = false;
			for(int j =  leftIndex; j < rightIndex - i ;j++){
				if(array[j].compareTo(array[j+1]) > 0){
					swap(array,j,j+1);
					flag = true;
				}
			}
			if(flag == false){
				break;
			}
		}
	}
	
	private void bidirectionalBubbleSort(T[] array,int leftIndex,int rightIndex){
		boolean flag = true;
		while(leftIndex < rightIndex && flag == true){
			flag = false;
			for(int i = leftIndex; i < rightIndex ; i++){
				if(array[i].compareTo(array[i +1]) > 0){
					swap(array,i,i+1);
					flag = true;
				}
			}
			rightIndex--;
			flag = false;
			for(int i = rightIndex;i > leftIndex;i--){
				if(array[i].compareTo(array[i-1]) < 0){
					swap(array,i,i-1);
					flag = true;
				}
			}
			leftIndex++;
		}
	}
	
	private void recursiveBubbleSort(T[] array,int leftIndex,int rightIndex,boolean flag){
		if(flag == true && leftIndex < rightIndex){
			for(int i = leftIndex; i < rightIndex;i++){
				if(array[i].compareTo(array[i+1]) > 0){
					swap(array,i,i+1);
					flag = true;
				}
			}
			rightIndex --;
			flag = false;
			for(int i = rightIndex;i > leftIndex;i--){
				if(array[i].compareTo(array[i-1]) < 0){
					swap(array,i,i-1);
					flag = true;
				}
			}
			leftIndex++;
			recursiveBubbleSort(array, leftIndex, rightIndex, flag);
		}
	}
	
	private void swap(T[] array, int i , int j){
		T aux = array[i];
		array[i] = array[j];
		array[j] = aux;
	}
	private boolean validateParams(T[] array, int leftIndex,int rightIndex){
		if(array == null || leftIndex >= rightIndex || rightIndex > array.length -1 || leftIndex < 0 ){
			return false;
		}else{
			return true;
		}
	}
}

File: Python/BubbleSort.py
#coding: utf-8
from Sort import Sort

__author__ = "Ionesio Junior"

class BubbleSort(Sort):
    ''' BubbleSort class implementation have many ways to sort using bubble sort idea '''

    def sort(self,lista,leftIndex = 0,rightIndex = None):
        ''' Generic Sort Method
            algorithm.sort(array) -> return  complete ordered array
            algorithm.sort(array,leftLimit,rightLimit) -> return partially ordered array
        '''        
        if rightIndex == None:
            rightIndex = len(lista) - 1
        
        if self.__validateParams(lista,leftIndex,rightIndex):
            #self.__simpleBubbleSort(lista,leftIndex,rightIndex)
            #self.__bidirectionalBubbleSort(lista,leftIndex,rightIndex)
            self.__recursiveBidirectionalBubbleSort(lista,leftIndex,rightIndex,True)
    

    def __simpleBubbleSort(self,lista,leftIndex,rightIndex):
        '''Simple Iterative Bubble Sort'''
        flag = True
        while(flag):
            flag = False
            for i in range(leftIndex,rightIndex):
                if(lista[i] > lista[i+1]):
                    lista[i] , lista[i +1] = lista[i+1] , lista[i]
                    flag = True
    
    def __bidirectionalBubbleSort(self,lista,leftIndex,rightIndex):
        ''' Iterative bidirectional bubble sort implementation in place'''
        flag = True
        while(flag  and leftIndex < rightIndex):
            flag = False;
            for i in range(leftIndex,rightIndex):
                if(lista[i] > lista[i+1]):
                    lista[i],lista[i+1] = lista[i+1] , lista[i]
                    flag = True
            flag = False
            rightIndex = rightIndex- 1
            for i in range(rightIndex,leftIndex +1,-1):
                if(lista[i] < lista[i - 1]):
                    lista[i-1],lista[i] = lista[i] , lista[i-1]
                    flag = True
            leftIndex = leftIndex +1

    def __recursiveBidirectionalBubbleSort(self,lista,leftIndex,rightIndex,flag):
        ''' Recursive bidirectional bubble sort implementation in place'''
        if(flag and leftIndex < rightIndex):
            flag = False
            for i in range(leftIndex,rightIndex):
                if(lista[i] > lista[i+1]):
                    lista[i] ,lista[i + 1] = lista[i+1] ,lista[i]
                    flag = True
            flag = False
            rightIndex = rightIndex -1
            for i in range(rightIndex,leftIndex,-1):
                if(lista[i] < lista[i-1]):
                    lista[i] , lista[i - 1] = lista[i-1],lista[i]
                    flag = True
            leftindex = leftIndex +1
            self.__recursiveBidirectionalBubbleSort(lista,leftIndex,rightIndex,flag)
    

    def __validateParams(self,lista , leftIndex , rightIndex):
        '''' This method verify if parameters are valid if some parameter is not valid the method returns false '''        
        if(lista == None  or leftIndex < 0 or rightIndex > len(lista) - 1 or leftIndex >= rightIndex):
            return False
        else:
            return True
